### PR TITLE
feat!(selection): Make IdMatches select filter OR composable

### PR DESCRIPTION
**Pull request by badosu**
Previously IdMatches would compose as AND, e.g.:

`_IdMatches_armpw_IdMatches_armflea` would keep only units that have the unitdef name armpw AND armflea at the same time, which was silly.

This commit makes it so that the previous filter keeps only units that have unitdef name armpw OR armflea.

Respectively `_Not_IdMatches_armpw_Not_IdMatches_armflea` maintains previous behavior, keep only units that don't have unitdef name armpw AND don't have unitdef name armflea.

Fixes https://github.com/beyond-all-reason/spring/issues/1058

Tested 5000 units selection filter in 4 different unit names with different compositions, didn't notice perceivable latency.

### CODE REVIEW

File: rts/Game/UI/SelectionKeyHandler.cpp [hunk 1]
```diff
@@ -383,6 +379,16 @@ void CSelectionKeyHandler::DoSelection(std::string selectString)
 			filter = ReadToken(selectString);
 		}
 
+		// We have to use this ad-hoc logic for handling OR-able IdMatches filtering
+		if (filter == "IdMatches") {
+			ReadDelimiter(selectString);
+
+			auto& matchSet = _not ? idNotMatchesSet : idMatchesSet;
```
Comment (sprunk) [inline rts/Game/UI/SelectionKeyHandler.cpp:386]: would it be possible to use `if (filter == "IdMatches" && not _not)` to avoid the `idNotMatchesSet` var?

File: rts/Game/UI/SelectionKeyHandler.cpp [hunk 2]
```diff
@@ -416,6 +422,32 @@ void CSelectionKeyHandler::DoSelection(std::string selectString)
 		}
 	}
 
+	if (!idMatchesSet.empty()) {
```
Comment (sprunk) [inline rts/Game/UI/SelectionKeyHandler.cpp:425]:
```suggestion
	/* Positive IdMatches use OR instead of AND,
	 * because that is intuitive and it's not possible
	 * for a unit to match two names anyway. */
	if (!idMatchesSet.empty()) {
```

File: rts/Game/UI/SelectionKeyHandler.cpp (hunk 1 above)
Comment (badosu) [reply rts/Game/UI/SelectionKeyHandler.cpp:386]: Sure, but how would we handle the `_Not_IdMatches` case?

Comment (sprunk) [reply rts/Game/UI/SelectionKeyHandler.cpp:386]: doesn't it keep the previous behaviour of AND if we don't handle it in this new block?

Comment (badosu) [reply rts/Game/UI/SelectionKeyHandler.cpp:386]: We'd have to restore the `IdMatches` filter singleton, it's a possibility. As someone having to deal with this code, would you prefer to have the whole IdMatches thing in its encapsulated logic or would you be fine with having IdMatches singleton that only works with _Not + the ad hoc case?

Comment (badosu) [reply rts/Game/UI/SelectionKeyHandler.cpp]: Good call!

Comment (sprunk) [reply rts/Game/UI/SelectionKeyHandler.cpp]: I think overall it would be good to keep explicit code only for cases that differ from the default behaviour (i.e. handle OR explicitly and let AND be done via the singleton). I don't know though, your call.

**Review by sprunk — APPROVED** [review]
Looks good, undraft?

**Comment by badosu** [comment]
When I'm back I'll do, also it's my understanding we're still not out of the merge freeze?

File: rts/Game/UI/SelectionKeyHandler.cpp [hunk 3]
```diff
@@ -367,6 +368,8 @@ void CSelectionKeyHandler::DoSelection(std::string selectString)
 
 	ReadDelimiter(selectString);
 
+	idMatchesSet.clear();
```
Comment (sprunk) [inline rts/Game/UI/SelectionKeyHandler.cpp:371]: Why a static var and not just a regular var on the stack declared here?

Comment (badosu) [reply rts/Game/UI/SelectionKeyHandler.cpp:371]: Good point, the closure is unnecessary. I think previous iteration used the scope.

**Comment by jamiewall** [comment]
Hi, sorry I'm not a github person, (owtfox) in Discord. Do you know when this will be put into the game? Quite excited to half the length of 90% of my filters! Also - I hope you don't mind - is this place where we might suggest specific things? I've always thought that adding customparams.techlevel would be a great filter, as I don't think NameContains, Category or RulesParam work! Thanks for everything you do :)

**Comment by sprunk** [comment]
There is no schedule but feel free to fix the outstanding issues (merge conflict, var member -> local) if you want it faster.

**Comment by jamiewall** [comment]
Okay, well thanks for responding so soon! This _might_ fix the class member/regular variable issue? So says Chat GPT :)

```
void CSelectionKeyHandler::DoSelection(std::string selectString) {
    std::set<std::string> idMatchesSet;  // Declare idMatchesSet as a local variable

    ReadDelimiter(selectString);

    while (true) {
        std::string filter = ReadDelimiter(selectString);

        // ... (existing code)

        /* Positive IdMatches use OR instead of AND,
         * because that is intuitive and it's not possible
         * for a unit to match two names anyway. */
        if (filter == "IdMatches" && !_not) {
            ReadDelimiter(selectString);

            idMatchesSet.insert(ReadToken(selectString));

            continue;
        }

        // ... (existing code)
    }

    if (!idMatchesSet.empty()) {
        // ... (existing code)
    }
}
```

I have never used Git before and this is the third comment I've ever made and the only 'coding' I ever did was on MySpace 16 years ago. So, I have no idea on the 'merge conflict', sorry. If you wouldn't mind helping with whatever that is?

**Comment by sprunk** [comment]
Fundamentally it means the upstream code has changed since the patch was created and some sort of "do something about line X which is located near line Y" cannot be resolved because there is no line X or Y anymore. In this case it's because the running changelog was wiped. It would be fine to get rid of the changelog entry from the patch altogether since that is managed centrally now.

**Comment by badosu** [comment]
I'll address this as soon as I can.

File: rts/Game/UI/SelectionKeyHandler.cpp [hunk 4]
```diff
@@ -389,6 +389,8 @@ void CSelectionKeyHandler::DoSelection(std::string selectString)
...
```
Comment (sprunk) [inline rts/Game/UI/SelectionKeyHandler.cpp:392]:
```suggestion
	std::unordered_set<std::string> idMatchesSet;
```

File: rts/Game/UI/SelectionKeyHandler.cpp [hunk 5]
```diff
@@ -24,6 +22,7 @@
 #include "System/UnorderedSet.hpp"
 
 CSelectionKeyHandler selectionKeys;
+std::unordered_set<std::string> idMatchesSet;
```
[the remark this answers is not in the document]
Comment (badosu) [reply rts/Game/UI/SelectionKeyHandler.cpp:25]: We don't have to allocate every invocation right? Although we could make it static I guess inside

[the remark this answers is not in the document]
Comment (sprunk) [reply rts/Game/UI/SelectionKeyHandler.cpp:25]: I think the perf differences are irrelevant and I'm suggesting a local because globals/statics make maintenance harder (which in this case isn't going to be terribly relevant either, but just on a principle)

**Comment by badosu** [comment]
Thanks for the help @sprunk

**Comment by sprunk** [comment]
Have you tested? I remember you asking somebody for help in that area but I sort of forgot to volunteer

**Comment by badosu** [comment]
Yeah, I was using the wrong build folder for the binary lol

**Comment by pandaro** [comment]
first look seem ok, you need to sets something in particular?

**Comment by badosu** [comment]
Nothing, all good now. The test scenarios I was most concerned about were:

- only 1 token in positive case
- only 1 token in negative case
- 1 token in both cases for different units
- 1 token in both cases for same units 
- multiple tokens testing different scenarios

I'm confident on the code and the tests I made, however there could be some edge case I didn't look for.

Thank you for testing!

**Comment by jamiewall** [comment]
I can't get this to work. Written 7 unique selection filters, ranging from 4 unit IDs through to 24 (and over) unit IDs. This is using Visible+ and SelectAll+ either side of the filter.

**Comment by sprunk** [comment]
Make sure you're testing on a fresh engine version. In particular the version of the engine on the live BAR server does not yet have this.

**Comment by jamiewall** [comment]
Thanks, Sprung. Another case of my ignorance. TIL merged does not mean live. Investigated following another user's query on Discord. Looking forward to seeing it in live - thanks again!
